presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,271 @@
+                case LESS_THAN:
+                case EQUAL:
+                case IS_DISTINCT_FROM:
+                    return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(max, sourceType));
+            }
+        }
+
+        int lowerBoundComparison = compare(literalType, literalValue, minInTargetType);
+
+        if (lowerBoundComparison < 0) {
+            // smaller than minimum representable value
+            switch (operator) {
+                case NOT_EQUAL:
+                case GREATER_THAN:
+                case GREATER_THAN_OR_EQUAL:
+                    return trueIfNotNull(cast.getExpression());
+                case EQUAL:
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                    return falseIfNotNull(cast.getExpression());
+                case IS_DISTINCT_FROM:
+                    return TRUE_LITERAL;
+            }
+        }
+
+        if (lowerBoundComparison == 0) {
+            // equal to min representable value
+            switch (operator) {
+                case LESS_THAN:
+                    return falseIfNotNull(cast.getExpression());
+                case LESS_THAN_OR_EQUAL:
+                    return new ComparisonExpression(EQUAL, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+                case GREATER_THAN_OR_EQUAL:
+                    return trueIfNotNull(cast.getExpression());
+                case GREATER_THAN:
+                case EQUAL:
+                case IS_DISTINCT_FROM:
+                    return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(min, sourceType));
+            }
+        }
+
+        // otherwise, somewhere in the middle, so check if the literal is representable in the
+        // type of the source expression
+
+        Object literalInSourceType = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(literalType, sourceType),
+                session.toConnectorSession(),
+                literalValue);
+
+        Object rountripLiteral = functionInvoker.invoke(
+                metadata.getFunctionRegistry().getCoercion(sourceType, literalType),
+                session.toConnectorSession(),
+                literalInSourceType);
+
+        if (compare(literalType, literalValue, rountripLiteral) != 0) {
+            // literal is not representable in source type
+            switch (operator) {
+                case EQUAL:
+                    return falseIfNotNull(cast.getExpression());
+                case NOT_EQUAL:
+                    return trueIfNotNull(cast.getExpression());
+                case IS_DISTINCT_FROM:
+                    return TRUE_LITERAL;
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                    return new ComparisonExpression(LESS_THAN_OR_EQUAL, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
```

Comment:
Since EQUAL is not an option, was this supposed to say "LESS_THAN" here?

Reply:
I think this is broken. It was there because of an assumption that CAST always "truncated" the least-significant part (i.e. that it does a "floor cast"), but that's sadly not the case :(